### PR TITLE
chore: Bump OpenDAL to v0.27

### PR DESCRIPTION
**Pull request by Xuanwo**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Bump OpenDAL to v0.27

Changlogs could be found at https://docs.rs/opendal/0.27.2/opendal/docs/changelog/index.html

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

**Comment by Xuanwo** [comment]
By the way, the `object_store`'s usage of `backend` and `services` is confused. Maybe we need a refactor on that? `object_store::backend::oss` and `object_store::services::Oss` are different import path.

**Comment by Xuanwo** [comment]
I plan to bump to version of opendal first and then clean up the re-export.

**Comment by killme2008** [comment]
> By the way, the `object_store`'s usage of `backend` and `services` is confused. Maybe we need a refactor on that? `object_store::backend::oss` and `object_store::services::Oss` are different import path.

Good catch! Looks like it was introduced by #911. The code reviewer didn't find it out.

File: src/common/procedure/src/store/state_store.rs [hunk 1]
```diff
@@ -72,29 +70,37 @@ impl StateStore for ObjectStateStore {
...
+            .scan()
+            .await
+            .map_err(|e| Error::ListState {
+                path: path_string.clone(),
+                source: e,
+            })?;
+
         // Note that there is no guarantee about the order between files and dirs
         // at the same level.
         // See https://docs.rs/opendal/0.25.2/opendal/raw/struct.TopDownWalker.html#note
```
Comment (killme2008) [inline src/common/procedure/src/store/state_store.rs:86]: The comment is outdated. Can you delete it?

File: src/object-store/Cargo.toml [hunk 2]
```diff
@@ -8,7 +8,7 @@ license.workspace = true
 lru = "0.9"
 async-trait = "0.1"
 futures = { version = "0.3" }
-opendal = { version = "0.25.1", features = [
+opendal = { version = "0.27.2", features = [
```
Comment (killme2008) [inline src/object-store/Cargo.toml:11]: According to our code style, prefer `0.27`.

File: src/object-store/src/cache_policy.rs [hunk 3]
```diff
@@ -12,112 +12,184 @@
...
+            }
+            Err(err) if err.kind() == ErrorKind::ObjectNotFound => {
+                let (rp, reader) = self.inner.read(&path, args.clone()).await?;
+                let size = rp.clone().into_metadata().content_length();
+                let _ = self
+                    .cache
+                    .write(
+                        &cache_path,
+                        OpWrite::new(size),
+                        Box::new(ReadWrppaer(reader)),
```
Comment (killme2008) [inline src/object-store/src/cache_policy.rs:103]: Typo: Wrppaer -> Wrapper

File: src/object-store/src/cache_policy.rs [hunk 4]
```diff
@@ -12,112 +12,184 @@
...
                         }
-                        Err(_) => inner.read(&path, args).await,
+                        return Ok((rp, Box::new(reader) as output::Reader));
+                    }
+                    Err(_) => {
+                        return self
+                            .inner
+                            .read(&path, args)
+                            .await
+                            .map(|(rp, r)| (rp, Box::new(r) as output::Reader))
```
Comment (killme2008) [inline src/object-store/src/cache_policy.rs:124]: Can we extract this block into a small function? It is duplicated below.

File: src/object-store/src/cache_policy.rs [hunk 5]
```diff
@@ -12,112 +12,184 @@
...
+    fn blocking_list(&self, path: &str, args: OpList) -> Result<(RpList, Self::BlockingPager)> {
+        self.inner.blocking_list(path, args)
+    }
+
+    fn blocking_scan(&self, path: &str, args: OpScan) -> Result<(RpScan, Self::BlockingPager)> {
+        self.inner.blocking_scan(path, args)
+    }
+}
+
+/// Workaround for output::Read doesn't implement input::Read
```
Comment (killme2008) [inline src/object-store/src/cache_policy.rs:182]: Adds a `TODO:` prefix.

**Comment by Xuanwo** [comment]
All comments have been address, PTAL.

**Review by killme2008 — APPROVED** [review]
LGTM. But there are still some toml format issues.

**Comment by v0y4g3r** [comment]
`object-store` crate was originally introduced to make accessor to oss implementations pluggable. Maybe it's time to remove this crate and directly use opendal's `Operator` since we are unlikely to migrate to any other oss implementation.

**Pull request merged**